addons/mail/static/src/model/model_core.js

```diff
@@ -2,134 +2,304 @@
+            return method.call(this, ...args);
+        });
+    }
+}
 
-export {
-    registerClassPatchModel,
-    registerFieldPatchModel,
-    registerInstancePatchModel,
-    registerIdentifyingFieldsPatch,
-    registerNewModel,
-    registry,
-};
+/**
+ * Overrides the record methods of the model specified by the `modelName`.
+ *
+ * @param {string} modelName The name of the model to which to apply the patch.
+ * @param {Object} methods Record methods to be overriden.
+ */
+export function patchRecordMethods(modelName, recordMethods) {
+    if (!registry.has(modelName)) {
+        throw new Error(`Cannot patch record methods on model "${modelName}": model must be registered before being patched.`);
+    }
+    const definition = registry.get(modelName);
+    for (const [name, method] of Object.entries(recordMethods)) {
+        if (typeof method !== 'function') {
+            throw new Error(`Cannot patch record method "${name}" on model "${modelName}": "${name}" must be a function.`);
+        }
+        if (!definition.get('recordMethods').has(name)) {
+            throw new Error(`Cannot patch record method "${name}" on model "${modelName}": method does not exist yet.`);
+        }
+        const methodBeforePatch = definition.get('recordMethods').get(name);
+        definition.get('recordMethods').set(name, function (...args) {
+            this._super = methodBeforePatch;
+            return method.call(this, ...args);
+        });
+    }
+}
 
+/**
+ * @param {Object} jsonDefinition The JSON definition of the model to register.
+ * @param {Object} jsonDefinition.fields
+ * @param {string[]} jsonDefinition.identifyingFields
+ * @param {Object} jsonDefinition.lifecycleHooks
+ * @param {Object} jsonDefinition.modelGetters
+ * @param {Object} jsonDefinition.modelMethods
+ * @param {string} jsonDefinition.name
+ * @param {OnChanges[]} jsonDefinition.onChanges
+ * @param {Object} jsonDefinition.recordGetters
+ * @param {Object} jsonDefinition.recordMethods
+ */
+export function registerModel({ fields, identifyingFields, lifecycleHooks, modelGetters, modelMethods, name, onChanges, recordGetters, recordMethods }) {
+    if (!name) {
+        throw new Error("Model is lacking a name.");
+    }
+    if (!identifyingFields) {
+        throw new Error(`Cannot register model "${name}": definition is lacking identifying fields.`);
+    }
+    if (registry.has(name)) {
+        throw new Error(`Cannot register model "${name}": model has already been registered.`);
+    }
+    const definition = new Map();
+    definition.set('name', name);
+    definition.set('identifyingFields', identifyingFields);
+    definition.set('lifecycleHooks', new Map());
+    definition.set('modelMethods', new Map());
+    definition.set('modelGetters', new Map());
+    definition.set('recordMethods', new Map());
+    definition.set('recordGetters', new Map());
+    definition.set('fields', new Map());
+    definition.set('onChanges', []);
```

Comment:
You didn't use map constructor for a specific reason here?
It's a bit verbose to repeat  `definition.set` on every line.